kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java

```diff
@@ -55,13 +70,90 @@ public SchemaManager(
       Optional<String> kafkaDataFieldName,
       Optional<String> timestampPartitionFieldName,
       Optional<List<String>> clusteringFieldName) {
+    this(
+        schemaRetriever,
+        schemaConverter,
+        bigQuery,
+        kafkaKeyFieldName,
+        kafkaDataFieldName,
+        timestampPartitionFieldName,
+        clusteringFieldName,
+        false,
+        new ConcurrentHashMap<>(),
+        new ConcurrentHashMap<>(),
+        new ConcurrentHashMap<>());
+  }
+
+  private SchemaManager(
+      SchemaRetriever schemaRetriever,
+      SchemaConverter<com.google.cloud.bigquery.Schema> schemaConverter,
+      BigQuery bigQuery,
+      Optional<String> kafkaKeyFieldName,
+      Optional<String> kafkaDataFieldName,
+      Optional<String> timestampPartitionFieldName,
+      Optional<List<String>> clusteringFieldName,
+      boolean intermediateTables,
+      ConcurrentMap<TableId, Object> tableCreateLocks,
+      ConcurrentMap<TableId, Object> tableUpdateLocks,
+      ConcurrentMap<TableId, com.google.cloud.bigquery.Schema> schemaCache) {
     this.schemaRetriever = schemaRetriever;
     this.schemaConverter = schemaConverter;
     this.bigQuery = bigQuery;
     this.kafkaKeyFieldName = kafkaKeyFieldName;
     this.kafkaDataFieldName = kafkaDataFieldName;
     this.timestampPartitionFieldName = timestampPartitionFieldName;
     this.clusteringFieldName = clusteringFieldName;
+    this.intermediateTables = intermediateTables;
+    this.tableCreateLocks = tableCreateLocks;
+    this.tableUpdateLocks = tableUpdateLocks;
+    this.schemaCache = schemaCache;
+  }
+
+  public SchemaManager forIntermediateTables() {
+    return new SchemaManager(
+        schemaRetriever,
+        schemaConverter,
+        bigQuery,
+        kafkaKeyFieldName,
+        kafkaDataFieldName,
+        timestampPartitionFieldName,
+        clusteringFieldName,
+        true,
+        tableCreateLocks,
+        tableUpdateLocks,
+        schemaCache
+    );
+  }
+
+  /**
+   * Fetch the most recent schema for the given table, assuming it has been created and/or updated
+   * over the lifetime of this schema manager.
+   * @param table the table to fetch the schema for; may be null
+   * @return the latest schema for that table; may be null if the table does not exist or has not
+   * been created or updated by this schema manager
+   */
+  public com.google.cloud.bigquery.Schema cachedSchema(TableId table) {
+    return schemaCache.get(table);
+  }
+
+  /**
+   * Create a new table in BigQuery, if it doesn't already exist. Otherwise, update the existing
+   * table to use the most-current schema.
+   * @param table The BigQuery table to create,
+   * @param topic The Kafka topic used to determine the schema.
+   */
+  public void createOrUpdateTable(TableId table, String topic) {
```

Comment:
Are we guaranteed that tasks on different workers won't try to create/update the same table?

Reply:
No, we aren't. This is mostly just a best-effort optimization to prevent lots of small batches from causing tasks to spam BigQuery with DDL requests. We should catch the case where the table create request fails because it already exists, though, and then respond by attempting a table update instead. Good catch, thanks!